File: scripts/md_to_pdf.mjs
// Generic markdown -> PDF renderer via Playwright (Chromium).
//
// Usage:  node scripts/md_to_pdf.mjs <input.md> <output.pdf>
//
// Renders the input markdown to a Sibley-Creek-styled HTML page, then
// uses Playwright's print-to-PDF to produce the output. Inline markdown
// parser is intentionally minimal — handles the subset of markdown used
// in onboarding/team docs (headings, paragraphs, lists, bold, italics,
// inline code, code fences, horizontal rules, links). For richer docs,
// swap in `marked` or `markdown-it`.

import { chromium } from "playwright";
import { readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";

const [, , inputArg, outputArg] = process.argv;
if (!inputArg || !outputArg) {
  console.error("usage: node scripts/md_to_pdf.mjs <input.md> <output.pdf>");
  process.exit(1);
}

const inputPath = resolve(process.cwd(), inputArg);
const outputPath = resolve(process.cwd(), outputArg);

const md = await readFile(inputPath, "utf8");

// --- minimal markdown -> HTML --------------------------------------------

function escapeHtml(s) {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function inline(s) {
  // links [text](url)
  s = s.replace(/\[([^\]]+)\]\(([^)]+)\)/g, (_, text, url) => {
    return `<a href="${escapeHtml(url)}">${inline(text)}</a>`;
  });
  // bold **text** (must come before italic)
  s = s.replace(/\*\*([^*]+)\*\*/g, "<strong>$1</strong>");
  // italic *text* or _text_
  s = s.replace(/(?<!\*)\*([^*]+)\*(?!\*)/g, "<em>$1</em>");
  // inline code `text`
  s = s.replace(/`([^`]+)`/g, "<code>$1</code>");
  return s;
}

function mdToHtml(src) {
  const lines = src.split(/\r?\n/);
  const out = [];
  let i = 0;
  let inList = false;

  function closeList() {
    if (inList) {
      out.push("</ol>");
      inList = false;
    }
  }

  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();

    if (trimmed === "") {
      closeList();
      i++;
      continue;
    }

    if (trimmed === "---") {
      closeList();
      out.push("<hr>");
      i++;
      continue;
    }

    // headings
    const h = trimmed.match(/^(#{1,4})\s+(.*)$/);
    if (h) {
      closeList();
      const level = h[1].length;
      out.push(`<h${level}>${inline(escapeHtml(h[2]))}</h${level}>`);
      i++;
      continue;
    }

    // numbered list items
    const ol = trimmed.match(/^\d+\.\s+(.*)$/);
    if (ol) {
      if (!inList) {
        out.push('<ol>');
        inList = true;
      }
      // gather indented continuation lines (sublists, paragraphs)
      let item = ol[1];
      let j = i + 1;
      while (j < lines.length && /^\s{3,}\S/.test(lines[j])) {
        item += " " + lines[j].trim();
        j++;
      }
      out.push(`<li>${inline(escapeHtml(item))}</li>`);
      i = j;
      continue;
    }

    // bullet list items
    const ul = trimmed.match(/^[-*]\s+(.*)$/);
    if (ul) {
      if (!inList) {
        out.push('<ul>');
        inList = true;
      }
      out.push(`<li>${inline(escapeHtml(ul[1]))}</li>`);
      i++;
      continue;
    }

    // plain paragraph (collect consecutive non-empty lines)
    closeList();
    const para = [];
    while (
      i < lines.length &&
      lines[i].trim() !== "" &&
      !lines[i].trim().startsWith("#") &&
      !lines[i].trim().startsWith("---") &&
      !/^\d+\.\s/.test(lines[i].trim()) &&
      !/^[-*]\s/.test(lines[i].trim())
    ) {
      para.push(lines[i].trim());
      i++;
    }
    out.push(`<p>${inline(escapeHtml(para.join(" ")))}</p>`);
  }

  closeList();
  return out.join("\n");
}

const body = mdToHtml(md);

const html = `<!doctype html>
<html lang="en-CA">
<head>
<meta charset="utf-8">
<title>Sibley Creek</title>
<style>
  @page { size: Letter; margin: 0.75in 0.85in; }
  html { font-family: "Manrope", "Helvetica Neue", Helvetica, Arial, sans-serif; color: #000; }
  body { font-size: 11pt; line-height: 1.55; max-width: 100%; }
  h1 { font-size: 22pt; font-weight: 800; letter-spacing: -0.01em; line-height: 1.15; margin: 0 0 12pt; border-bottom: 1px solid #000; padding-bottom: 8pt; }
  h2 { font-size: 13pt; font-weight: 800; letter-spacing: -0.005em; margin: 18pt 0 6pt; }
  h3 { font-size: 11pt; font-weight: 700; margin: 12pt 0 4pt; text-transform: none; }
  p { margin: 0 0 8pt; }
  ol, ul { margin: 0 0 10pt 0; padding-left: 20pt; }
  li { margin-bottom: 4pt; }
  li > strong:first-child { font-weight: 700; }
  hr { border: 0; border-top: 1px solid #000; margin: 16pt 0; }
  code { font-family: "IBM Plex Mono", "SFMono-Regular", Consolas, monospace; font-size: 10pt; background: #f3f3f3; padding: 1pt 4pt; border-radius: 2pt; }
  strong { font-weight: 700; }
  em { font-style: italic; }
  a { color: #000; text-decoration: underline; }
  /* Tight first-element margin so heading sits at top */
  h1:first-child { margin-top: 0; }
</style>
</head>
<body>
${body}
</body>
</html>`;

// --- render via Playwright -----------------------------------------------

const browser = await chromium.launch();
const ctx = await browser.newContext();
const page = await ctx.newPage();
await page.setContent(html, { waitUntil: "networkidle" });
await page.pdf({
  path: outputPath,
  format: "Letter",
  printBackground: true,
  margin: { top: "0.75in", bottom: "0.75in", left: "0.85in", right: "0.85in" },
});
await browser.close();

console.log(`wrote: ${outputPath}`);

File: claude-ref/research/_md_to_pdf.py
"""Convert a markdown file to PDF using Python-Markdown + Chrome headless."""
import subprocess
import sys
from pathlib import Path
import markdown

if len(sys.argv) != 3:
    print("Usage: py _md_to_pdf.py <input.md> <output.pdf>")
    sys.exit(1)

md_path = Path(sys.argv[1]).resolve()
pdf_path = Path(sys.argv[2]).resolve()
html_path = pdf_path.with_suffix(".html")

md_text = md_path.read_text(encoding="utf-8")
html_body = markdown.markdown(md_text, extensions=["tables", "fenced_code", "toc"])

css = """
@page { size: Letter; margin: 0.75in; }
body { font-family: Georgia, 'Times New Roman', serif; font-size: 11pt; line-height: 1.5; color: #1a1a1a; max-width: 7in; }
h1 { font-size: 22pt; margin-top: 0; border-bottom: 2px solid #1a1a1a; padding-bottom: 6pt; }
h2 { font-size: 16pt; margin-top: 24pt; border-bottom: 1px solid #888; padding-bottom: 3pt; }
h3 { font-size: 13pt; margin-top: 18pt; }
h4 { font-size: 11pt; margin-top: 12pt; font-weight: bold; }
p { margin: 6pt 0; }
ul, ol { margin: 6pt 0; padding-left: 24pt; }
li { margin: 3pt 0; }
strong { font-weight: bold; }
em { font-style: italic; }
table { border-collapse: collapse; margin: 12pt 0; font-size: 10pt; width: 100%; }
th, td { border: 1px solid #888; padding: 4pt 8pt; text-align: left; vertical-align: top; }
th { background-color: #eaeaea; font-weight: bold; }
code { font-family: 'Consolas', 'Courier New', monospace; background-color: #f4f4f4; padding: 1pt 3pt; font-size: 9.5pt; }
pre { background-color: #f4f4f4; padding: 8pt; border-radius: 3pt; overflow-x: auto; font-size: 9.5pt; }
blockquote { border-left: 3px solid #888; padding-left: 12pt; color: #444; margin: 12pt 0; font-style: italic; }
hr { border: 0; border-top: 1px solid #ccc; margin: 18pt 0; }
a { color: #1a1a1a; text-decoration: underline; }
"""

html_full = f"""<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>{md_path.stem}</title><style>{css}</style></head>
<body>{html_body}</body></html>"""

html_path.write_text(html_full, encoding="utf-8")

chrome_path = r"C:\Program Files\Google\Chrome\Application\chrome.exe"
cmd = [
    chrome_path,
    "--headless",
    "--disable-gpu",
    "--no-pdf-header-footer",
    f"--print-to-pdf={pdf_path}",
    f"file:///{html_path.as_posix()}",
]
result = subprocess.run(cmd, capture_output=True, text=True)
if result.returncode != 0:
    print("Chrome stderr:", result.stderr)
    sys.exit(result.returncode)
print(f"PDF written: {pdf_path}")
